File: Java/Language/Basics/src/SavingsAccount.java
package banking;

final class SavingsAccount extends Account implements Profitable {

	//non-reassignable field
	static final double MIN_BAL = 5000;

	SavingsAccount() {
		balance = MIN_BAL;
	}

	public void deposit(double amount) {
		balance += amount;
	}

	public void withdraw(double amount) throws InsufficientFundsException {
		if(balance - amount < MIN_BAL)
			throw new InsufficientFundsException();
		balance -= amount;
	}

	public double interest(int months) {
		float rate = balance < 4 * MIN_BAL ? MIN_RATE : MIN_RATE + 1;
		return balance * rate * months / 1200;
	}
}

File: CSharp/Basics/AbstractClassTest/Banking/SavingsAccount.cs
namespace Banking
{
	sealed class SavingsAccount : Account
	{
		const decimal MinBal = 5000;

		internal SavingsAccount()
		{
			Balance = MinBal;
		}

		public override void Deposit(decimal amount)
		{
			Balance += amount;
		}
		
		public override void Withdraw(decimal amount)
		{
			if(Balance - amount < MinBal)
				throw new InsufficientFundsException();
			Balance -= amount;
		}
	}
}

